File: apps/macos/Kelpie/Handlers/InteractionHandler.swift
import WebKit

private struct TapExecution {
    let requestedX: Double
    let requestedY: Double
    let appliedX: Double
    let appliedY: Double
    let offsetX: Double
    let offsetY: Double

    var responsePayload: [String: Any] {
        [
            "x": requestedX,
            "y": requestedY,
            "appliedX": appliedX,
            "appliedY": appliedY,
            "offsetX": offsetX,
            "offsetY": offsetY
        ]
    }
}

/// Handles click, tap, fill, type, selectOption, check, uncheck.
struct InteractionHandler {
    let context: HandlerContext

    func register(on router: Router) {
        router.register("click") { body in await click(body) }
        router.register("tap") { body in await tap(body) }
        router.register("fill") { body in await fill(body) }
        router.register("type") { body in await typeText(body) }
        router.register("select-option") { body in await selectOption(body) }
        router.register("check") { body in await setChecked(body, checked: true) }
        router.register("uncheck") { body in await setChecked(body, checked: false) }
    }

    @MainActor
    private func click(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        guard let selector = body["selector"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "selector is required")
        }
        let color = overlayColor(from: body)
        do {
            let result = try await context.evaluateJSReturningJSON(selectorActivationScript(selector), tabId: tabId)
            let diagnostics = result["diagnostics"] as? [String: Any]
            if result.isEmpty || result["error"] as? String == "not_found" {
                return errorResponse(
                    code: "ELEMENT_NOT_FOUND",
                    message: "Element not found: \(selector)",
                    diagnostics: diagnostics
                )
            }
            if result["error"] as? String == "not_visible" {
                return errorResponse(
                    code: "ELEMENT_NOT_VISIBLE",
                    message: "Element is not visible or is obscured: \(selector)",
                    diagnostics: diagnostics
                )
            }
            if let center = result["center"] as? [String: Any],
               let x = double(center["x"]), let y = double(center["y"]) {
                await context.showTouchIndicator(x: x, y: y, color: color, tabId: tabId)
            } else {
                await context.showTouchIndicatorForElement(selector, color: color, tabId: tabId)
            }
            return successResponse(["element": result])
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func tap(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        guard var requestedX = double(body["x"]), var requestedY = double(body["y"]) else {
            return errorResponse(code: "MISSING_PARAM", message: "x and y are required")
        }
        // Convert screenshot image-pixel coordinates to CSS viewport pixels.
        // Screenshots are captured at native resolution (points * backingScaleFactor),
        // but elementFromPoint expects CSS viewport pixels. LLMs read pixel coords
        // from the screenshot image, so "screenshot" is the natural input space.
        let coordinateSpace = (body["coordinateSpace"] as? String ?? "viewport").lowercased()
        if coordinateSpace == "screenshot" {
            do {
                let metrics = try await context.screenshotViewportMetrics(tabId: tabId)
                let dpr = max(metrics.devicePixelRatio, 1)
                requestedX /= dpr
                requestedY /= dpr
            } catch {
                if let tabError = tabErrorResponse(from: error) { return tabError }
                return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
            }
        }
        let execution: TapExecution
        do {
            execution = try await calibratedTapExecution(
                requestedX: requestedX,
                requestedY: requestedY,
                calibration: TapCalibrationStore.current(),
                tabId: tabId
            )
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
        await context.showTouchIndicator(
            x: execution.appliedX,
            y: execution.appliedY,
            color: overlayColor(from: body),
            tabId: tabId
        )
        do {
            let diagnostics = try await context.evaluateJSReturningJSON(tapScript(execution: execution), tabId: tabId)
            return successResponse(execution.responsePayload.merging(["diagnostics": diagnostics]) { _, new in new })
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func fill(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        guard let selector = body["selector"] as? String, let value = body["value"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "selector and value are required")
        }
        let color = overlayColor(from: body)
        let mode = (body["mode"] as? String ?? "instant").lowercased()
        let delay = body["delay"] as? Int ?? 50

        if mode == "typing" {
            let focusJS = "document.querySelector('\(JSEscape.string(selector))')?.focus()"
            _ = try? await context.evaluateJS(focusJS, tabId: tabId)
            do {
                let focusResult = try await context.evaluateJSReturningJSON(
                    fillElementScript(selector: selector, value: ""),
                    tabId: tabId
                )
                let diagnostics = focusResult["diagnostics"] as? [String: Any]
                if focusResult.isEmpty || focusResult["error"] as? String == "not_found" {
                    return errorResponse(
                        code: "ELEMENT_NOT_FOUND",
                        message: "Element not found: \(selector)",
                        diagnostics: diagnostics
                    )
                }
                if focusResult["error"] as? String == "not_editable" {
                    return errorResponse(
                        code: "INVALID_PARAMS",
                        message: "Element is not an editable form control: \(selector)",
                        diagnostics: diagnostics
                    )
                }
                return await typeText([
                    "selector": selector,
                    "text": value,
                    "delay": delay,
                    "color": body["color"] as Any,
                    "tabId": tabId as Any
                ].compactMapValues { $0 })
            } catch {
                if let tabError = tabErrorResponse(from: error) { return tabError }
                return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
            }
        }

        do {
            let result = try await context.evaluateJSReturningJSON(fillElementScript(selector: selector, value: value), tabId: tabId)
            let diagnostics = result["diagnostics"] as? [String: Any]
            if result.isEmpty || result["error"] as? String == "not_found" {
                return errorResponse(
                    code: "ELEMENT_NOT_FOUND",
                    message: "Element not found: \(selector)",
                    diagnostics: diagnostics
                )
            }
            if result["error"] as? String == "not_editable" {
                return errorResponse(
                    code: "INVALID_PARAMS",
                    message: "Element is not an editable form control: \(selector)",
                    diagnostics: diagnostics
                )
            }
            await context.showTouchIndicatorForElement(selector, color: color, tabId: tabId)
            return successResponse(result)
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func typeText(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        guard let text = body["text"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "text is required")
        }
        let color = overlayColor(from: body)
        if let selector = body["selector"] as? String {
            let focusJS = "document.querySelector('\(JSEscape.string(selector))')?.focus()"
            _ = try? await context.evaluateJS(focusJS, tabId: tabId)
            await context.showTouchIndicatorForElement(selector, color: color, tabId: tabId)
        }
        let delay = body["delay"] as? Int ?? 50
        for char in text {
            if context.scriptPlaybackState?.isAbortRequested() == true { break }
            let escapedChar = JSEscape.string(String(char))
            let charJS = """
            (function() {
                \(formControlMutationScript())
                var el = document.activeElement;
                if (!el) return;
                el.dispatchEvent(new KeyboardEvent('keydown', {key: '\(escapedChar)', bubbles: true}));
                el.dispatchEvent(new KeyboardEvent('keypress', {key: '\(escapedChar)', bubbles: true}));
                kelpieWriteFormControlValue(el, kelpieReadFormControlValue(el) + '\(escapedChar)');
                kelpieDispatchFormControlInput(el);
                el.dispatchEvent(new KeyboardEvent('keyup', {key: '\(escapedChar)', bubbles: true}));
            })()
            """
            _ = try? await context.evaluateJS(charJS, tabId: tabId)
            try? await Task.sleep(nanoseconds: UInt64(delay) * 1_000_000)
        }
        let finalizeJS = """
        (function() {
            \(formControlMutationScript())
            var el = document.activeElement;
            if (!el) return;
            kelpieDispatchFormControlChange(el);
        })()
        """
        _ = try? await context.evaluateJS(finalizeJS, tabId: tabId)
        return successResponse(["typed": text])
    }

    @MainActor
    private func selectOption(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        guard let selector = body["selector"] as? String, let value = body["value"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "selector and value are required")
        }
        let js = """
        (function() {
            var el = document.querySelector('\(JSEscape.string(selector))');
            if (!el) return null;
            el.value = '\(JSEscape.string(value))';
            el.dispatchEvent(new Event('change', {bubbles: true}));
            var opt = el.options?.[el.selectedIndex];
            return {selected: {value: el.value, text: opt ? opt.text : el.value}};
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js, tabId: tabId)
            if result.isEmpty { return errorResponse(code: "ELEMENT_NOT_FOUND", message: "Element not found: \(selector)") }
            await context.showTouchIndicatorForElement(selector, color: overlayColor(from: body), tabId: tabId)
            return successResponse(result)
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func setChecked(_ body: [String: Any], checked: Bool) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        guard let selector = body["selector"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "selector is required")
        }
        let js = """
        (function() {
            var el = document.querySelector('\(JSEscape.string(selector))');
            if (!el) return null;
            el.checked = \(checked);
            el.dispatchEvent(new Event('change', {bubbles: true}));
            return {checked: el.checked};
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js, tabId: tabId)
            if result.isEmpty { return errorResponse(code: "ELEMENT_NOT_FOUND", message: "Element not found: \(selector)") }
            await context.showTouchIndicatorForElement(selector, color: overlayColor(from: body), tabId: tabId)
            return successResponse(result)
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    private func overlayColor(from body: [String: Any]) -> String {
        HandlerContext.hexToRGB(body["color"] as? String ?? "#3B82F6")
    }

    private func double(_ value: Any?) -> Double? {
        (value as? NSNumber)?.doubleValue
    }

    @MainActor
    private func calibratedTapExecution(
        requestedX: Double,
        requestedY: Double,
        calibration: TapCalibration,
        tabId: String?
    ) async throws -> TapExecution {
        let viewport = try await viewportSize(tabId: tabId)
        let appliedX = clamp(requestedX + calibration.offsetX, min: 0, max: max(viewport.width - 1, 0))
        let appliedY = clamp(requestedY + calibration.offsetY, min: 0, max: max(viewport.height - 1, 0))
        return TapExecution(
            requestedX: requestedX,
            requestedY: requestedY,
            appliedX: appliedX,
            appliedY: appliedY,
            offsetX: calibration.offsetX,
            offsetY: calibration.offsetY
        )
    }

    @MainActor
    private func viewportSize(tabId: String?) async throws -> (width: Double, height: Double) {
        let result = try await context.evaluateJSReturningJSON("""
        (function() {
            return {
                width: Math.max(window.innerWidth || 0, 1),
                height: Math.max(window.innerHeight || 0, 1)
            };
        })()
        """, tabId: tabId)
        let width = double(result["width"]) ?? 1
        let height = double(result["height"]) ?? 1
        return (width, height)
    }

    private func clamp(_ value: Double, min lower: Double, max upper: Double) -> Double {
        guard lower <= upper else { return lower }
        return Swift.min(Swift.max(value, lower), upper)
    }

    private func tapScript(execution: TapExecution) -> String {
        """
        (function() {
            \(interactionHelpersScript())
            var requestedX = \(execution.requestedX);
            var requestedY = \(execution.requestedY);
            var appliedX = \(execution.appliedX);
            var appliedY = \(execution.appliedY);
            var offsetX = \(execution.offsetX);
            var offsetY = \(execution.offsetY);
            var hook = window.__kelpieTapCalibration;
            if (hook && typeof hook.onAutomationTap === 'function') {
                try {
                    hook.onAutomationTap({
                        requestedX: requestedX,
                        requestedY: requestedY,
                        appliedX: appliedX,
                        appliedY: appliedY,
                        offsetX: offsetX,
                        offsetY: offsetY
                    });
                } catch (error) {}
            }
            var eventTarget = document.elementFromPoint(appliedX, appliedY) || document.body || document.documentElement;
            if (!eventTarget) {
                return kelpieTapDiagnostics(null, requestedX, requestedY, appliedX, appliedY, offsetX, offsetY);
            }
            if (typeof eventTarget.focus === 'function') {
                try { eventTarget.focus({preventScroll: true}); } catch (error) { try { eventTarget.focus(); } catch (focusError) {} }
            }
            function dispatchMouse(type, button, buttons) {
                eventTarget.dispatchEvent(new MouseEvent(type, {
                    bubbles: true,
                    cancelable: true,
                    composed: true,
                    clientX: appliedX,
                    clientY: appliedY,
                    screenX: appliedX,
                    screenY: appliedY,
                    detail: type === 'click' ? 1 : 0,
                    button: button,
                    buttons: buttons
                }));
            }
            function dispatchPointer(type, button, buttons) {
                if (typeof window.PointerEvent !== 'function') {
                    return;
                }
                eventTarget.dispatchEvent(new PointerEvent(type, {
                    bubbles: true,
                    cancelable: true,
                    composed: true,
                    clientX: appliedX,
                    clientY: appliedY,
                    screenX: appliedX,
                    screenY: appliedY,
                    pointerId: 1,
                    pointerType: 'touch',
                    isPrimary: true,
                    button: button,
                    buttons: buttons
                }));
            }
            dispatchPointer('pointerdown', 0, 1);
            dispatchMouse('mousedown', 0, 1);
            dispatchPointer('pointerup', 0, 0);
            dispatchMouse('mouseup', 0, 0);
            if (typeof eventTarget.click === 'function') {
                eventTarget.click();
            } else {
                dispatchMouse('click', 0, 0);
            }
            return kelpieTapDiagnostics(eventTarget, requestedX, requestedY, appliedX, appliedY, offsetX, offsetY);
        })()
        """
    }
}

File: apps/android/app/src/main/java/com/kelpie/browser/handlers/InteractionHandler.kt
package com.kelpie.browser.handlers

import com.kelpie.browser.network.Router
import com.kelpie.browser.network.errorResponse
import com.kelpie.browser.network.successResponse

class InteractionHandler(
    private val ctx: HandlerContext,
) {
    fun register(router: Router) {
        router.register("click") { click(it) }
        router.register("tap") { tap(it) }
        router.register("fill") { fill(it) }
        router.register("type") { type(it) }
        router.register("select-option") { selectOption(it) }
        router.register("check") { check(it) }
        router.register("uncheck") { uncheck(it) }
    }

    private suspend fun click(body: Map<String, Any?>): Map<String, Any?> {
        val selector = body["selector"] as? String ?: return errorResponse("MISSING_PARAM", "selector is required")
        return try {
            val result = ctx.evaluateJSReturningJSON(selectorActivationScript(selector))
            val diagnostics = result["diagnostics"] as? Map<String, Any?>
            when (result["error"]) {
                "not_found" -> {
                    errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector", diagnostics)
                }

                "not_visible" -> {
                    errorResponse("ELEMENT_NOT_VISIBLE", "Element is not visible or is obscured: $selector", diagnostics)
                }

                null -> {
                    val center = result["center"] as? Map<*, *>
                    val x = (center?.get("x") as? Number)?.toDouble()
                    val y = (center?.get("y") as? Number)?.toDouble()
                    if (x != null && y != null) {
                        ctx.showTouchIndicator(x, y, ctx.overlayColor(body))
                    } else {
                        ctx.showTouchIndicatorForElement(selector, ctx.overlayColor(body))
                    }
                    successResponse(mapOf("element" to result))
                }

                else -> {
                    errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector")
                }
            }
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun tap(body: Map<String, Any?>): Map<String, Any?> {
        var requestedX = (body["x"] as? Number)?.toDouble() ?: return errorResponse("MISSING_PARAM", "x and y are required")
        var requestedY = (body["y"] as? Number)?.toDouble() ?: return errorResponse("MISSING_PARAM", "x and y are required")
        val coordinateSpace = (body["coordinateSpace"] as? String ?: "viewport").lowercase()
        if (coordinateSpace == "screenshot") {
            try {
                val metrics = ctx.viewportMetrics()
                val dpr = maxOf(metrics.devicePixelRatio, 1.0)
                requestedX /= dpr
                requestedY /= dpr
            } catch (e: Exception) {
                return errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
            }
        }
        val execution =
            try {
                calibratedTapExecution(requestedX, requestedY, TapCalibrationStore.current())
            } catch (e: Exception) {
                return errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
            }
        ctx.showTouchIndicator(execution.appliedX, execution.appliedY, ctx.overlayColor(body))
        return try {
            val diagnostics = ctx.evaluateJSReturningJSON(tapScript(execution))
            successResponse(
                mapOf(
                    "x" to execution.requestedX,
                    "y" to execution.requestedY,
                    "appliedX" to execution.appliedX,
                    "appliedY" to execution.appliedY,
                    "offsetX" to execution.offsetX,
                    "offsetY" to execution.offsetY,
                    "diagnostics" to diagnostics,
                ),
            )
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun fill(body: Map<String, Any?>): Map<String, Any?> {
        val selector = body["selector"] as? String ?: return errorResponse("MISSING_PARAM", "selector is required")
        val value = body["value"] as? String ?: return errorResponse("MISSING_PARAM", "value is required")
        val mode = (body["mode"] as? String ?: "instant").lowercase()
        val delay = (body["delay"] as? Int) ?: 50
        if (mode == "typing") {
            return try {
                val focusResult = ctx.evaluateJSReturningJSON(fillElementScript(selector, ""))
                val diagnostics = focusResult["diagnostics"] as? Map<String, Any?>
                when (focusResult["error"]) {
                    "not_found" -> errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector", diagnostics)
                    "not_editable" -> errorResponse("INVALID_PARAMS", "Element is not an editable form control: $selector", diagnostics)
                    null ->
                        type(
                            mapOf(
                                "selector" to selector,
                                "text" to value,
                                "delay" to delay,
                                "color" to body["color"],
                            ),
                        )
                    else -> errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector", diagnostics)
                }
            } catch (e: Exception) {
                errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
            }
        }
        return try {
            val result = ctx.evaluateJSReturningJSON(fillElementScript(selector, value))
            val diagnostics = result["diagnostics"] as? Map<String, Any?>
            when (result["error"]) {
                "not_found" -> errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector", diagnostics)
                "not_editable" -> errorResponse("INVALID_PARAMS", "Element is not an editable form control: $selector", diagnostics)
                null -> {
                    ctx.showTouchIndicatorForElement(selector, ctx.overlayColor(body))
                    // Canonical fill shape (matches iOS/macOS and FillResponse in
                    // shared/api-types.ts): top-level {selector, value, element}.
                    // Spreading the script result preserves the diagnostics-style
                    // `element` field returned by fillElementScript.
                    successResponse(result)
                }

                else -> {
                    errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector", diagnostics)
                }
            }
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun type(body: Map<String, Any?>): Map<String, Any?> {
        val text = body["text"] as? String ?: return errorResponse("MISSING_PARAM", "text is required")
        val selector = body["selector"] as? String
        if (selector != null) {
            val focusJs = "document.querySelector('${JSEscape.string(selector)}')?.focus()"
            runCatching { ctx.evaluateJS(focusJs) }
            ctx.showTouchIndicatorForElement(selector, ctx.overlayColor(body))
        }
        val delayMs = (body["delay"] as? Int) ?: 50
        return try {
            for (char in text) {
                if (ctx.scriptPlaybackState?.isAbortRequested() == true) break
                val escapedChar = JSEscape.string(char.toString())
                val charJs =
                    """
                    (function() {
                        ${formControlMutationScript()}
                        var el = document.activeElement;
                        if (!el) return;
                        el.dispatchEvent(new KeyboardEvent('keydown', {key: '$escapedChar', bubbles: true}));
                        el.dispatchEvent(new KeyboardEvent('keypress', {key: '$escapedChar', bubbles: true}));
                        kelpieWriteFormControlValue(el, kelpieReadFormControlValue(el) + '$escapedChar');
                        kelpieDispatchFormControlInput(el);
                        el.dispatchEvent(new KeyboardEvent('keyup', {key: '$escapedChar', bubbles: true}));
                    })()
                    """.trimIndent()
                ctx.evaluateJS(charJs)
                kotlinx.coroutines.delay(delayMs.toLong())
            }
            ctx.evaluateJS(
                """
                (function() {
                    ${formControlMutationScript()}
                    var el = document.activeElement;
                    if (!el) return;
                    kelpieDispatchFormControlChange(el);
                })()
                """.trimIndent(),
            )
            successResponse(mapOf("typed" to text))
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun selectOption(body: Map<String, Any?>): Map<String, Any?> {
        val selector = body["selector"] as? String ?: return errorResponse("MISSING_PARAM", "selector is required")
        val value = body["value"] as? String ?: return errorResponse("MISSING_PARAM", "value is required")
        // Canonical select-option shape (matches iOS/macOS and SelectOptionResponse
        // in shared/api-types.ts): top-level {selected: {value, text}}.
        val js =
            "(function(){var el=document.querySelector('${JSEscape.string(selector)}');" +
                "if(!el)return null;el.value='${JSEscape.string(value)}';" +
                "el.dispatchEvent(new Event('change',{bubbles:true}));" +
                "var opt=el.options?el.options[el.selectedIndex]:null;" +
                "return{selected:{value:el.value,text:opt?opt.text:el.value}};})()"
        return try {
            val result = ctx.evaluateJSReturningJSON(js)
            if (result.isEmpty()) {
                errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector")
            } else {
                ctx.showTouchIndicatorForElement(selector, ctx.overlayColor(body))
                successResponse(result)
            }
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun check(body: Map<String, Any?>): Map<String, Any?> = setChecked(body, true)

    private suspend fun uncheck(body: Map<String, Any?>): Map<String, Any?> = setChecked(body, false)

    private suspend fun setChecked(
        body: Map<String, Any?>,
        checked: Boolean,
    ): Map<String, Any?> {
        val selector = body["selector"] as? String ?: return errorResponse("MISSING_PARAM", "selector is required")
        // Canonical check/uncheck shape (matches iOS/macOS and CheckResponse
        // in shared/api-types.ts): top-level {checked}.
        val js =
            "(function(){var el=document.querySelector('${JSEscape.string(selector)}');" +
                "if(!el)return null;el.checked=$checked;" +
                "el.dispatchEvent(new Event('change',{bubbles:true}));" +
                "return{checked:el.checked};})()"
        return try {
            val result = ctx.evaluateJSReturningJSON(js)
            if (result.isEmpty()) {
                errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector")
            } else {
                ctx.showTouchIndicatorForElement(selector, ctx.overlayColor(body))
                successResponse(result)
            }
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun calibratedTapExecution(
        requestedX: Double,
        requestedY: Double,
        calibration: TapCalibration,
    ): TapExecution {
        val viewport = viewportSize()
        val appliedX = clamp(requestedX + calibration.offsetX, 0.0, maxOf(viewport.first - 1.0, 0.0))
        val appliedY = clamp(requestedY + calibration.offsetY, 0.0, maxOf(viewport.second - 1.0, 0.0))
        return TapExecution(
            requestedX = requestedX,
            requestedY = requestedY,
            appliedX = appliedX,
            appliedY = appliedY,
            offsetX = calibration.offsetX,
            offsetY = calibration.offsetY,
        )
    }

    private suspend fun viewportSize(): Pair<Double, Double> {
        val result =
            ctx.evaluateJSReturningJSON(
                """
                (function() {
                    return {
                        width: Math.max(window.innerWidth || 0, 1),
                        height: Math.max(window.innerHeight || 0, 1)
                    };
                })()
                """.trimIndent(),
            )
        val width = (result["width"] as? Number)?.toDouble() ?: 1.0
        val height = (result["height"] as? Number)?.toDouble() ?: 1.0
        return width to height
    }

    private fun clamp(
        value: Double,
        lower: Double,
        upper: Double,
    ): Double =
        when {
            lower > upper -> lower
            value < lower -> lower
            value > upper -> upper
            else -> value
        }

    private fun tapScript(execution: TapExecution): String =
        """
        (function() {
            ${interactionHelpersScript()}
            var requestedX = ${execution.requestedX};
            var requestedY = ${execution.requestedY};
            var appliedX = ${execution.appliedX};
            var appliedY = ${execution.appliedY};
            var offsetX = ${execution.offsetX};
            var offsetY = ${execution.offsetY};
            var hook = window.__kelpieTapCalibration;
            if (hook && typeof hook.onAutomationTap === 'function') {
                try {
                    hook.onAutomationTap({
                        requestedX: requestedX,
                        requestedY: requestedY,
                        appliedX: appliedX,
                        appliedY: appliedY,
                        offsetX: offsetX,
                        offsetY: offsetY
                    });
                } catch (error) {}
            }
            var eventTarget = document.elementFromPoint(appliedX, appliedY) || document.body || document.documentElement;
            if (!eventTarget) {
                return kelpieTapDiagnostics(null, requestedX, requestedY, appliedX, appliedY, offsetX, offsetY);
            }
            if (typeof eventTarget.focus === 'function') {
                try { eventTarget.focus({preventScroll: true}); } catch (error) { try { eventTarget.focus(); } catch (focusError) {} }
            }
            function dispatchMouse(type, button, buttons) {
                eventTarget.dispatchEvent(new MouseEvent(type, {
                    bubbles: true,
                    cancelable: true,
                    composed: true,
                    clientX: appliedX,
                    clientY: appliedY,
                    screenX: appliedX,
                    screenY: appliedY,
                    detail: type === 'click' ? 1 : 0,
                    button: button,
                    buttons: buttons
                }));
            }
            function dispatchPointer(type, button, buttons) {
                if (typeof window.PointerEvent !== 'function') {
                    return;
                }
                eventTarget.dispatchEvent(new PointerEvent(type, {
                    bubbles: true,
                    cancelable: true,
                    composed: true,
                    clientX: appliedX,
                    clientY: appliedY,
                    screenX: appliedX,
                    screenY: appliedY,
                    pointerId: 1,
                    pointerType: 'touch',
                    isPrimary: true,
                    button: button,
                    buttons: buttons
                }));
            }
            dispatchPointer('pointerdown', 0, 1);
            dispatchMouse('mousedown', 0, 1);
            dispatchPointer('pointerup', 0, 0);
            dispatchMouse('mouseup', 0, 0);
            if (typeof eventTarget.click === 'function') {
                eventTarget.click();
            } else {
                dispatchMouse('click', 0, 0);
            }
            return kelpieTapDiagnostics(eventTarget, requestedX, requestedY, appliedX, appliedY, offsetX, offsetY);
        })();
        """.trimIndent()
}
